File: app/src/main/java/com/stupidtree/hichat/ui/widgets/EmoticonsEditText.java
package com.stupidtree.hichat.ui.widgets;

import android.content.Context;
import android.graphics.Bitmap;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.style.ImageSpan;
import android.util.AttributeSet;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.stupidtree.hichat.utils.ImageUtils;
import com.stupidtree.hichat.utils.TextUtils;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static com.stupidtree.hichat.ui.widgets.EmoticonsTextView.buildPattern;

public class EmoticonsEditText extends androidx.appcompat.widget.AppCompatEditText {


    public EmoticonsEditText(@NonNull Context context) {
        super(context);
    }

    public EmoticonsEditText(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public EmoticonsEditText(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }



    @Override
    public void setText(CharSequence text, BufferType type) {
        if (!TextUtils.isEmpty(text)) {
            super.setText(replace(text.toString()), type);
        } else {
            super.setText(text, type);
        }
    }


    @Override
    public void append(CharSequence text, int start, int end) {
        if(!TextUtils.isEmpty(text)){
            super.append(replace(text.toString()), start, end);
        }else{
            super.append(text, start, end);
        }

    }

    private CharSequence replace(String text) {
        try {
            SpannableString spannableString = new SpannableString(text);
            int start = 0;
            Pattern pattern = buildPattern();
            Matcher matcher = pattern.matcher(text);
            while (matcher.find()) {
                String faceText = matcher.group();
                String key = faceText.substring(1);
                key = key.substring(key.indexOf("[") + 1, key.indexOf("]"));
                int imageId = getContext().getResources().getIdentifier("yunmoji_" + key, "drawable", getContext().getPackageName());
                Bitmap bitmap = ImageUtils.getBitmapFromDrawable(getContext(), imageId,getTextSize()*1.2f);
                ImageSpan imageSpan = new ImageSpan(getContext(), bitmap);
                int startIndex = text.indexOf(faceText, start);
                int endIndex = startIndex + faceText.length();
                if (startIndex >= 0)
                    spannableString.setSpan(imageSpan, startIndex, endIndex, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                start = (endIndex - 1);
            }
            return spannableString;
        } catch (Exception e) {
            return text;
        }
    }
}

File: app/src/main/java/com/stupidtree/cloudliter/ui/widgets/EmoticonsEditText.kt
package com.stupidtree.cloudliter.ui.widgets

import android.content.Context
import android.text.Spannable
import android.text.SpannableString
import android.text.style.ImageSpan
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatEditText
import com.stupidtree.cloudliter.utils.ImageUtils
import com.stupidtree.cloudliter.utils.TextUtils

class EmoticonsEditText : AppCompatEditText {
    constructor(context: Context) : super(context) {}
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {}
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    override fun setText(text: CharSequence, type: BufferType) {
        if (!TextUtils.isEmpty(text)) {
            super.setText(replace(text.toString()), type)
        } else {
            super.setText(text, type)
        }
    }

    override fun append(text: CharSequence, start: Int, end: Int) {
        if (!TextUtils.isEmpty(text)) {
            super.append(replace(text.toString()), start, end)
        } else {
            super.append(text, start, end)
        }
    }

    private fun replace(text: String): CharSequence {
        return try {
            val spannableString = SpannableString(text)
            var start = 0
            val pattern = EmoticonsTextView.buildPattern()
            val matcher = pattern.matcher(text)
            while (matcher.find()) {
                val faceText = matcher.group()
                var key = faceText.substring(1)
                key = key.substring(key.indexOf("[") + 1, key.indexOf("]"))
                val imageId = context.resources.getIdentifier("yunmoji_$key", "drawable", context.packageName)
                val bitmap = ImageUtils.getBitmapFromDrawable(context, imageId, textSize * 1.3f)
                val imageSpan = ImageSpan(context, bitmap)
                val startIndex = text.indexOf(faceText, start)
                val endIndex = startIndex + faceText.length
                if (startIndex >= 0) spannableString.setSpan(imageSpan, startIndex, endIndex, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
                start = endIndex - 1
            }
            spannableString
        } catch (e: Exception) {
            text
        }
    }
}